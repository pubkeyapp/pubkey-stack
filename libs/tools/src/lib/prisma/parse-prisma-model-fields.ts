export function parsePrismaModelFields(fields: string[]): ParsedPrismaModelField[] {
  return fields
    .map((field) => (field.includes(':') ? field : `${field}: String`))
    .map((field) => {
      const [name, type, attrs] = field.split(':')
      const isOptional = type.includes('?')

      return {
        name,
        type: type.replace('?', ''),
        isOptional,
        attributes: attrs?.length
          ? attrs.split('@').map((attr) => {
              const [name, params] = attr.split('(')
              return {
                name,
                params: params?.split(',').map((param) => {
                  const [name] = param.split(')')
                  return { name }
                }),
              }
            })
          : [],
      }
    })
}

export interface ParsedPrismaModelField {
  name: string
  type: string
  isOptional?: boolean
  attributes?: { name: string; params?: { name: string; params?: string[] }[] }[]
}
