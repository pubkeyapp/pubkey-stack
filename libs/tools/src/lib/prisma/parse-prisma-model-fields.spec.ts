import { parsePrismaModelFields } from './parse-prisma-model-fields'

describe('parse-prisma-model-fields', () => {
  it.each([
    { fields: ['description:String', 'amount:Int'] },
    { fields: ['description:String?'] },
    { fields: ['owner:User', 'ownerId:String'] },
    { fields: ['owner:User?', 'ownerId:String?'] },
    {
      fields: [
        'title',
        'content',
        'views:Int',
        'owner:User?:relation@(fields: [ownerId], references: [id])',
        'ownerId:String?',
      ],
    },
  ])(`should normalize model input with name (%o)`, ({ fields }) => {
    const parsed = parsePrismaModelFields(fields)

    expect(parsed).toMatchSnapshot()
  })
})
