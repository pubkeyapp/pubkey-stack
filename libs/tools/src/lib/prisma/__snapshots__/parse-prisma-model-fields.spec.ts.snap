// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse-prisma-model-fields should normalize model input with name ({
  fields: [
    'title',
    'content',
    'views:Int',
    'owner:User?:relation@(fields: [ownerId], references: [id])',
    'ownerId:String?',
    [length]: 5
  ]
}) 1`] = `
[
  {
    "attributes": [],
    "isOptional": false,
    "name": "title",
    "type": " String",
  },
  {
    "attributes": [],
    "isOptional": false,
    "name": "content",
    "type": " String",
  },
  {
    "attributes": [],
    "isOptional": false,
    "name": "views",
    "type": "Int",
  },
  {
    "attributes": [
      {
        "name": "relation",
        "params": undefined,
      },
      {
        "name": "",
        "params": [
          {
            "name": "fields",
          },
        ],
      },
    ],
    "isOptional": true,
    "name": "owner",
    "type": "User",
  },
  {
    "attributes": [],
    "isOptional": true,
    "name": "ownerId",
    "type": "String",
  },
]
`;

exports[`parse-prisma-model-fields should normalize model input with name ({ fields: [ 'description:String', 'amount:Int', [length]: 2 ] }) 1`] = `
[
  {
    "attributes": [],
    "isOptional": false,
    "name": "description",
    "type": "String",
  },
  {
    "attributes": [],
    "isOptional": false,
    "name": "amount",
    "type": "Int",
  },
]
`;

exports[`parse-prisma-model-fields should normalize model input with name ({ fields: [ 'description:String?', [length]: 1 ] }) 1`] = `
[
  {
    "attributes": [],
    "isOptional": true,
    "name": "description",
    "type": "String",
  },
]
`;

exports[`parse-prisma-model-fields should normalize model input with name ({ fields: [ 'owner:User', 'ownerId:String', [length]: 2 ] }) 1`] = `
[
  {
    "attributes": [],
    "isOptional": false,
    "name": "owner",
    "type": "User",
  },
  {
    "attributes": [],
    "isOptional": false,
    "name": "ownerId",
    "type": "String",
  },
]
`;

exports[`parse-prisma-model-fields should normalize model input with name ({ fields: [ 'owner:User?', 'ownerId:String?', [length]: 2 ] }) 1`] = `
[
  {
    "attributes": [],
    "isOptional": true,
    "name": "owner",
    "type": "User",
  },
  {
    "attributes": [],
    "isOptional": true,
    "name": "ownerId",
    "type": "String",
  },
]
`;
